### DISCUSSION TITLE
include image output of LaTeX blocks in non-LaTeX formats

### DISCUSSION THREAD

**Discussion by wuqui** [quarto-dev/quarto-cli#4293] [Q&A] [answered]
I think it would be very powerful if one could include include the output of LaTeX blocks as images in html formats. I could imagine several use cases (e.g. complex tables), but the use case I’m interested most involves drawing trees in LaTeX using the forest package like this:

``````
```{=tex}
\begin{forest}
  [1
    [2]
    [3]
  ]
\end{forest}
```
``````

These work great for rendering to pdf, but when exporting to html (e.g. revealjs), I would like Quarto/pandoc to include a standalone image of these blocks.

Is this possible out-of-the-box already somehow? I have searched for quite a while, but maybe I missed it.

If not, any ideas how I could make it work – maybe using a Lua filter? I have no idea where to start for this.

**Comment by cderv** [comment]
> Is this possible out-of-the-box already somehow?

This is not possible for now in Quarto. However, using external tools you can have this to work. 

One example is using R in the qmd document. This will allow you to use this R package https://yonicd.github.io/texPreview/ that will do just what you are looking for. It should work as is within Quarto document as it works with **knitr** already. 

> maybe using a Lua filter? I have no idea where to start for this.

Using the syntax you described would indeed require to develop a Lua filter that would do quite the same as the R package I mentioned. This could be a good Quarto extensions. I don't think there is one yet. 

We have some information on creating extensions and working with Lua in our doc: https://quarto.org/docs/extensions/lua.html

Hope it helps

**Comment by wuqui** [comment, marked as answer]
Thanks a lot for your quick and helpful reply.

I have now checked both options and both work. After getting texPreview to work, I had a sense that a Quarto extension would be the more elegant solution, though, so I tried to get that running and eventually did. I tried to make it more broadly applicable by reading LaTeX requirements from a `preamble.sty` file, so that it’s not bound to my particular use case of the forest library and so that requirements are shared between Quarto’s ‘normal’ pdf processing and the extra runs for creating svg images.

I’m new to pandoc/Quarto and Lua, so suggestions are most welcome: https://github.com/wuqui/teximg. If it’s interesting to others, it could also be added to the Quarto extension list – I don’t know whether it’s useful/fancy enough for that, though …

One thing I haven’t figured out yet is how to specify the size of the image?

**Reply by cderv** [reply]
Awesome well done ! 

> If it’s interesting to others, it could also be added to the Quarto extension list – I don’t know whether it’s useful/fancy enough for that, though …

I think that depends on how much you plan to use it and maintain this. This is something I wanted to do too, but did not had time for it. That is cool you did it. I would be happy to take over also at some point if you don't plan to maintain it yourself. 

Regarding listing, you should add some tags to your github repo if you want discoverability (like `quarto`, `quarto-extension`, ...). And if you want it to be listed on the website, you can [do a PR](https://github.com/quarto-dev/quarto-web/tree/main/docs/extensions/listings)

Thanks a lot for this !

**Reply by wuqui** [reply]
Thank you very much, I’m happy to hear that 🙂

I would be happy to have it listed. My use case involving the forest package is probably quite rare, but I think the general functionality of including standalone images of LaTeX blocks in non-LaTeX outputs could be more widely useful, so I would be happy if it is useful for others.

Thanks for the tip about the PR, I didn’t know that. I think I’ll do that then. Before I do, I would just like to see if I can optimize two things:

1. How could I specify the display (font) size of the svg file (for example, in html output)? It’s quite small.
2. Is it possible to use the LaTeX requirements specified in the Quarto yaml? That would seem more Quartonic than relying on a preamble.sty 😉

**Reply by cderv** [reply]
> I think the general functionality of including standalone images of LaTeX blocks in non-LaTeX outputs could be more widely useful, so I would be happy if it is useful for others.

Definitely useful from my experience !

> Is it possible to use the LaTeX requirements specified in the Quarto yaml? That would seem more Quartonic than relying on a preamble.sty

If some dependencies are required by your extensions, best things is to add the latex dependency you need using this Quarto API from Lua https://quarto.org/docs/extensions/lua-api.html#dependencies. See example in our extensions like how we allow [using fontawesome incons in LaTeX document](https://github.com/quarto-ext/fontawesome/blob/5aa16bb67ff8d26ce9a972bd8184a0005ec95971/_extensions/fontawesome/fontawesome.lua#L1-L3)

Otherwise, this is dependency need by the user to run each raw TeX block right ? 

The idea of `preamble.sty` is nice. Otherwise, you can add some config in YAML for your extension. See example with our [latex-environment extension](https://github.com/quarto-ext/latex-environment) - it can read some information from YAML config. 

If you want something more scoped to the raw tex block, you'll need to leverage other structure because you can't set ID or attributes on those structure. So it would need to be like enclosing the raw block in a div with attributes or leveraging another syntax than the raw block, but more complex filter at the end.

**Comment by jdutant** [comment]
I'm writing a Quarto/Pandoc filter that does this, with pretty extensive customization options, e.g. zoom that you can specify for the whole document, classes or individual elements. It can also embed svg images within the html output file. It's functional, though I've focused on HTML output and MacOS/Linux only so far. I'll do further tests and add a better sample document before I create a release. But you can already have a look (issues and suggestions welcome!)

https://github.com/dialoa/imagify

Note: to see a sample output you'll have to clone and run `make generate && open test/expected.html`. The sample output in the [online documentation](https://dialoa.github.io/imagify/) doesn't yet work but thta's just because it's generated on GitHub using the [pandoc/latex docker image](https://hub.docker.com/r/pandoc/latex), which unfortunately doesn't include the standalone LaTeX document class.

**Reply by cderv** [reply]
Nice! Thanks for sharing ! 

@mcanouil another one to list on your list ! If not there already :)

**Reply by wuqui** [reply]
That’s looking very promising and much more elaborate and flexible than what I came up with 👍🏻

So I guess you’ll keep working on this and you’ll file a pull request for it to be added to the Quarto extension listings, right? I’m asking because I had planned to do that for the filter I made above, but if you want to go ahead with your solution, I won’t, because it looks like yours will be the way to go.

I didn’t have much time to give it a spin, but I quickly tested it and it didn’t work with including an additional LaTeX library. The output of the `forest` below wasn’t rendered. Do I need to specify the `\usepackage{forest}` some other way? PDF output works, but in html the forest isn’t imagified.

```
---
filters:
  - imagify
header-includes: \usepackage{forest}
---

::: imagify
$$\binom{n}{k} = \frac{n!}{k!(n-k)!}$$
:::

::: imagify
\begin{forest}
  [1
    [2]
    [3]
  ]
\end{forest}
:::
```

**Reply by jdutant** [reply]
Thanks! I plan to submit it to the Quarto list. But precisely I thought it needed better testing first 😄 

The issue here isn't with specific packages but something about how `header-includes` behaves in Quarto as opposed to Pandoc. If you save the above as `sample.qmd` and run:

```
pandoc -L imagify.lua -f markdown sample.qmd -o test.html
```

You get the desired result (provided the forest package is in your local LaTeX distribution). But if you run:

```
quarto render sample.qmd
```

The text `\usepackage{forest}` is included in the body text, and the imagify filter doesn't pick it up to create its images. Wrapping it in raw tags (better practice) takes it out of the body text, but doesn't help imagify either:

~~~
 header-includes: |
    ```{=latex}
    \usepackage{forest}
    ```
~~~

I'll dump the meta from within Quarto to figure out what's going on. Stay tuned.

**Reply by cderv** [reply]
With Quarto it will be better to use the YAML fields documented https://quarto.org/docs/output-formats/pdf-basics.html#latex-includes
For header, it will be `include-in-header` 

We have special processing to handle files and text within the same YAML field, and also to merge anything passed to it by the user with our own inclusion for Pandoc. 

I don't remember if `header-includes` still works though - Our doc may not be up to date. 

Regarding your specific extension, are you using this specific keys ? I see you may be (https://github.com/dialoa/imagify/blob/main/_extensions/imagify/imagify.lua#L692) 

It may be not easy to handle the headers from YAML into your lua filter. You could adopt an approach where you ask for the package to use in a imagify configuration, and for PDF output you pass it as includes to Pandoc from Lua directly : https://quarto.org/docs/extensions/lua-api.html#includes 

Also, the challenge we faced with such tool is : should LaTeX package be provided for all doc or per chunk you want to "imagify" ? 

Anyway, quick feedback - I didn't look to deep in what you have done but very cool from what I saw! Thank you

Hope it helps

**Reply by jdutant** [reply]
**SOLUTION**. Quarto erases LaTeX commands in `header-includes` when targeting HTML output. You must (1) provide it in the imagifier key instead, and (2) if you want to explicitly declare Raw latex, use a separate metadata file. That is, this will work:

`source.qmd`

``` markdown
---
filters:
  - imagify
format:
  html:
    imagify:
      header-includes: \usepackage{forest}
  latex:
    header-includes: \usepackage{forest}
  pdf:
    header-includes: \usepackage{forest}
---
(rest of source file)
```

Or simply (if you don't need LaTeX/PDF output):

``` markdown
---
filters:
  - imagify
imagify:
   header-includes: \usepackage{forest}
---
(rest of source file)
```

The PDF and imagified HTML render correcty with `quarto render source.md --to pdf,html`. 

However, (point 2), if for some reason you need to specify that your `header-includes` is raw LaTeX, you need to use a separate metadata file. This doesn't work:

```markdown
---
filters:
  - imagify
imagify:
      header-includes: |
         ```{=latex}
         \usepackage{forest}
         ```
---
(rest of source file)
```

Because in HTML output, the explicitly specified `{=latex}` element is simply erased. To use it you need to provide it as a separate file instead. This will work:

`source.md`
```markdown
---
format:
  html:
    metadata-files: 
    - imagify-meta.yaml
---
```

`imagify-meta.yaml`
```yaml
imagify:
  header-includes: |
    ```{=latex}
    \usepackage{forest}
    ```
```

Note (point 1 again) that even in a separate metadata file, you still can't use the root `header-includes` key. You have to use the one in `imagify`. This will not work:

`imagify-meta.yaml`
```yaml
header-includes: \usepackage{forest}
```

Nor this:

~~~yaml
header-includes: |
  ```{=latex}
  \usepackage{forest}
  ```
~~~

For in HTML output, Quarto duplicates the header-includes as raw html. Imagifying works but your document starts with the text "\usepackage{forest}" or "\'\'\'{=latex}\\n\\usepackage{forest}\'\'\'\\\n" . See this extract of Quarto's meta element dump for the first and the second:

```
header-includes: table: 0x6000036fde40
  1: table: 0x6000036fedc0
    1: RawInline (Format "tex") "\\usepackage{forest}"
  2: table: 0x6000036fedc0
    1: RawBlock (Format "html") "\\usepackage{forest}"
```

```
header-includes: table: 0x600003111f40
  1: table: 0x600003112ec0
    1: RawBlock (Format "latex") "\\usepackage{forest}"
  2: table: 0x600003112ec0
    1: RawBlock (Format "html") "```{=latex}\n\\usepackage{forest}\n```\n"
```

I'm not sure why Quarto does this!

**LONGER TERM SOLUTION**. Not convenient! I'm thinking of providing a `latex-packages` option (to be placed in root, imagify or imagify classes) as an alternative. 

```yaml
latex-packages:
- forest
- name: circ
  options: physics
```

Looks like a neat metadata addition anyway: the filter could supply the relevant `header-includes` in PDF output as well.

**Reply by jdutant** [reply]
@cderv Thanks! Sorry our messages crossed.

"the challenge we faced with such tool is : should LaTeX package be provided for all doc or per chunk you want to "imagify" ?"

The filter provides both, as we've had both use cases in the journal Dialectica. You can specify packages for all imagified elements (in the `imagify` key), or just for those that appear in a Div of a certain class. You can even specify them on the attributes of a single Div. See [documentation](https://dialoa.github.io/imagify/#options-reference) (still crude).

"You could adopt an approach where you ask for the package to use in a imagify configuration, and for PDF output you pass it as includes to Pandoc from Lua directly". 

👍 Indeed, the filter could by default push 'up' to the root whatever `header-includes` is provided under the global `imagify` key (or the global and classes?). 

"With Quarto it will be better to use the YAML fields documented https://quarto.org/docs/output-formats/pdf-basics.html#latex-includes For header, it will be include-in-header"

I think that won't work right? Because we're rendering the whole document to HTML, while we're turning its LaTeX elements to LaTeX/PDF. `in-header` can't be used, because it'll place "usepackage..." in the HTML document header. (The filter could look for those there and remove them, but that's a hassle.) `include-in-header` doesn't work either, because Quarto (surprisingly!) includes the `.tex` as RawInline with *html* format:

`source.md`
```markdown
---
...
include-in-header: test.tex
---
````

`test.tex`
```
\usepackage{forest}
```

Dump of Quarto's meta when rendering to html:
```
header-includes: table: 0x600002ce9e40
  1: table: 0x600002ceadc0
    1: RawBlock (Format "html") "\\usepackage{forest}"
```

It wouldn't help to specify `include-in-header` under `format: pdf:`, because the latter is erased when we target HTML output. (Perhaps I could pick it up by placing the filter before quarto, but it'd be better to find a solution that doesn't require that.)

Re LaTeX fields: "are you using this specific keys ? I see you may be (https://github.com/dialoa/imagify/blob/main/_extensions/imagify/imagify.lua#L692)"

Yes, the filter is [looking for all the official LaTeX-relevant Pandoc keys](https://dialoa.github.io/imagify/#pandocs-latex-options) that are applicable to a standalone document. From a quick test (dump of the meta element from within the filter) they are passed to the filter. The filter looks for them both at the root and under `imagify` (the latter overriding the former). Seems fine?

**Reply by cderv** [reply]
> 👍 Indeed, the filter could by default push 'up' to the root whatever header-includes is provided under the global imagify key (or the global and classes?).

What I meant really is not just pushing to the root, it is handling metadata under `imagify` key to do the right thing. Quarto has a specific Lua API to help, among other thing, add some dependencies to LaTeX or HTML. This means it does not rely on Pandoc handling of `header-includes`.  Put the link again at the doc: https://quarto.org/docs/extensions/lua-api.html#includes

I believe this would give you some more flexibility as you can catch the meta, do something and remove it for Pandoc to ignore it later on. 

> I think that won't work right? Because we're rendering the whole document to HTML, while we're turning its LaTeX elements to LaTeX/PDF. in-header can't be used, because it'll place "usepackage..." in the HTML document header. (The filter could look for those there and remove them, but that's a hassle.)

And the above is definitely for that without much hassle thanks to the Quarto Lua API. 

> include-in-header doesn't work either, because Quarto (surprisingly!) includes the .tex as RawInline with html format

It don't find that surprising. `include-in-header` is not really meant to be global as usually file included will differ in HTML and PDF. So for multi format document, one would add `include-in-header` under each format key. 

````yaml
format:
  html:
    include-in-header: head.html
  pdf:
    include-in-header: head.tex
````

`include-in-header` assumes that the file passed to it is compatible with the format used. 

> Yes, the filter is [looking for all the official LaTeX-relevant Pandoc keys](https://dialoa.github.io/imagify/#pandocs-latex-options) that are applicable to a standalone document.

Quarto has some custom fields so this may need to be adapted or at least be made compatible. Not so much difference I think. 

> The filter looks for them both at the root and under imagify (the latter overriding the former). Seems fine?

Yes - It seems fine to me to use same name key as default Pandoc. What I am suggesting is doing some special treatment in Lua and not just passing them through to override. This way I think you could catch a raw content in LaTeX that would be remove in HTML output and use it for your HTML processing. 

Anyhow, maybe I am missing something. If you are interesting, I can try do an example of what I mean in your repo. We can also discuss this topic in an issue in your repo. 

Thanks a lot for the discussion !

**Reply by jdutant** [reply]
Thanks @cderv this is helpful. I now get the point about the Lua API. I'll start posting issues I can think of for making use of all the quarto goodies, and to make sure the filter works well with the ecosystem. Any suggestions / PRs welcome, and I'll come here for help.

**Reply by jdutant** [reply]
One quick question though: in Quarto, can a filter retrieve what options (e.g. [pdf-engine](https://quarto.org/docs/reference/formats/pdf.html#format-options), etc) the user sets for LaTeX/PDF output when we're rendering a document to HTML? If I recall well from my tests meta values set at the root like `mainfont` are passed no matter what the output format is, but those set under `format: pdf:` are ditched. Can we still access them with the Lua api, e.g. quarto.doc.pdf_engine()?

**Reply by cderv** [reply]
> Can we still access them with the Lua api, e.g. quarto.doc.pdf_engine()?

That is a good question. you should try to see what it renders in Lua. It could be a matter of where we're at in the Lua evaluation order, but it is highly possible that for HTML all keys under `pdf` format at not read for HTML output. 🤔 

Please do ask your questions like that in new topics maybe so that we can discuss more clearly with the rest of the team specific use case like that.

**Reply by jdutant** [reply]
Thanks, I'll test it first.
